lib/ASTPrinter/CodePrinter.cpp

```diff
@@ -0,0 +1,210 @@
+#include "AST/Types.hpp"
+#include "ASTVisitor.hpp"
+#include "TypePrinter.hpp"
+
+#include <llvm/Support/raw_ostream.h>
+#include <string>
+
+namespace glu::ast {
+
+/// @brief CodePrinter is a visitor that converts AST nodes back to Glu source
+/// code.
+///
+/// This class generates valid Glu source code from AST nodes, primarily for
+/// decompilation of function prototypes and type declarations. It supports:
+/// - FunctionDecl (without body)
+/// - StructDecl
+/// - EnumDecl
+///
+/// The generated code should be syntactically valid Glu code that can be
+/// used as interface declarations.
+class CodePrinter : public ASTVisitor<CodePrinter> {
+    llvm::raw_ostream &_out; ///< The output stream to write the code to
+    TypePrinter _typePrinter; ///< Type printer for formatting types
+    size_t _indent = 0; ///< Current indentation level
+
+public:
+    /// @brief Constructs a CodePrinter object.
+    /// @param out The output stream to write the generated code to.
+    CodePrinter(llvm::raw_ostream &out = llvm::outs())
+        : _out(out), _typePrinter(true, true) // Use readable type names
+    {
+    }
+
+    /// @brief Visit a ModuleDecl and print its contents as Glu code
+    /// @param node The ModuleDecl node to print
+    void visitModuleDecl(ModuleDecl *node)
+    {
+        // Print each top-level declaration in the module
+        for (auto *decl : node->getDecls()) {
+            if (auto *funcDecl = llvm::dyn_cast<FunctionDecl>(decl)) {
+                visit(funcDecl);
+                _out << "\n";
+            } else if (auto *structDecl = llvm::dyn_cast<StructDecl>(decl)) {
+                visit(structDecl);
+                _out << "\n";
+            } else if (auto *enumDecl = llvm::dyn_cast<EnumDecl>(decl)) {
+                visit(enumDecl);
+                _out << "\n";
+            }
```

Comment:
you can just `visit(decl);` and print your newline, anything unhandled should just not be printed: `visitASTNode()` should do nothing